eos-boot-upgrade

```diff
@@ -0,0 +1,267 @@
+    for (dirpath, dirnames, filenames) in os.walk(dir_path):
+        for f in filenames:
+            fp = os.path.join(dirpath, f)
+            total_size += (os.path.getsize(fp) if os.path.isfile(fp) else 0)
+
+    return total_size
+
+def get_partition_free_space(dir_path):
+    '''Calculate the free space available in the partition'''
+    (_, _, free) = shutil.disk_usage(dir_path);
+    return free
+
+def check_esp_space(esp_mount, ostree_boot_efi_path):
+    '''Check that we have enough space in the ESP to copy the new data'''
+    ostree_boot_efi_size = get_dir_size(ostree_boot_efi_path);
+    esp_free_space = get_partition_free_space(esp_mount);
+
+    return esp_free_space > ostree_boot_efi_size
+
+def _dir_cmp(dcmp):
+    '''Helper for directories comparison'''
+    if dcmp.diff_files or dcmp.right_only:
+        return True
+    for sub_dcmp in dcmp.subdirs.values():
+        if _dir_cmp(sub_dcmp):
+            return True
+    return False
+
+def check_for_upgrade(esp_mount, ostree_boot_efi_path):
+    '''Recursively compare the files to upgrade'''
+    dcmp = filecmp.dircmp(esp_mount, ostree_boot_efi_path)
+    return _dir_cmp(dcmp)
+
+def do_uefi_upgrade(esp_mount, ostree_boot_efi_path):
+    '''Actually start the UEFI upgrade copying the files'''
+    tmp_dest_directory = os.path.join(esp_mount, 'new')
+
+    try:
+        distutils.dir_util.copy_tree(ostree_boot_efi_path, tmp_dest_directory)
+        upgrade_list = distutils.dir_util.copy_tree(ostree_boot_efi_path, esp_mount)
+        os.sync()
+    finally:
+        distutils.dir_util.remove_tree(tmp_dest_directory)
+
+    print('[*] Updated:')
+    print(*upgrade_list, sep='\n')
+
+def start_uefi_upgrade():
+    '''Start the UEFI files upgrade'''
+    esp_mount = '/boot/efi'
+    ostree_boot_efi_path = os.path.join(cmdline.options['ostree'], 'boot/efi')
+
+    if not os.path.isdir(esp_mount):
+        raise BootUpgradeException('[!] ESP mountpoint', esp_mount, 'does not exist')
+
+    try:
+        disk = get_disk();
+        esp_partition = get_esp(disk);
+
+        with mount(esp_partition.device, esp_mount) as esp_mount:
+            if not check_for_upgrade(esp_mount, ostree_boot_efi_path):
+                print('[*] No UEFI files to upgrade')
+                return
+
+            if not check_esp_space(esp_mount, ostree_boot_efi_path):
+                raise BootUpgradeException('[!] Not enough space in the ESP')
+
+            do_uefi_upgrade(esp_mount, ostree_boot_efi_path)
+
+    except (BootUpgradeException, subprocess.CalledProcessError) as BUException:
+        print(BUException)
+
+def start_conf_upgrade():
+    '''Start the GRUB configuration upgrade'''
+    conf_path = '/boot/grub/grub.cfg'
+    new_conf_path = '/usr/lib/grub/conf/grub.cfg'
+
+    try:
+        if filecmp.cmp(conf_path, new_conf_path):
+            print('[*] No configuration file to upgrade')
+            return
+
+        shutil.copyfile(new_conf_path, conf_path);
+        os.sync()
```

Comment:
I think it would be safer to copy the new conf to `/boot`, `fsync()` it, then `rename()` it over `conf_path`. This sequence is rumoured to guarantee that in the event of a crash `/boot/grub/grub.cfg` would either contain its old or its new contents. By copying directly over it, its contents might be truncated or filled with `NUL` bytes if the system crashes before `sync()` completes.

Reply:
`/boot` is on the same partition, so we can directly `rename()` it over.